design-proposals/dynamic-primary-pod-nic-name.md

```diff
@@ -0,0 +1,466 @@
+
+### Alternative 4 - Use the existing Downward API mechanism
+
+Currently, the VMI controller sets the `kubevirt.io/network-info` annotation on the virt-launcher pod in two cases:
+
+1. The VMI has an SR-IOV NIC.
+2. The VMI uses a network binding plugin
+
+The annotation's value is derived fom Multus' network-status annotation and is injected into the virt-launcher pod
+using K8s' [Downward API](https://kubernetes.io/docs/concepts/workloads/pods/downward-api/) mechanism.
+
+A new field will be added to the `NetworkInfo` struct:
+
+```go
+type NetworkInfo struct {
+Interfaces []Interface `json:"interfaces,omitempty"`
+PrimaryPodInterfaceName string `json:primaryPodInterfaceName,omitempty`
+}
+```
+
+A new mechanism should be added between virt-handler and virt-launcher to share this information, because:
+
+1. The primary pod interface name is also required by virt-handler
+2. The virt-handler does not have access to pod objects
+
+#### Pros
+
+#### Cons
+
+- Requires a new mechanism to have the information in virt-handler.
+- All virt-launcher pods will have to use the downward API mechanism instead of just the subset using it today
+
+## Specifying the primary pod interface name
+
+### Adding a cluster-wide preferred pod NIC name
+
+A new field will be added in KubeVirt's cluster-wide network configuration, which will specify the preferred pod
+interface name that should be bound to the VM interface destined for pod network:
+
+```go
+type NetworkConfiguration struct {
+...
+PreferredPodPrimaryInterfaceName string `json:"preferredPodPrimaryInterfaceName,omitempty"`
+...
+}
+```
+
+#### API Example
+
+```yaml
+apiVersion: kubevirt.io/v1
+kind: KubeVirt
+metadata:
+  name: kubevirt
+  namespace: kubevirt
+spec:
+  configuration:
+    network:
+      preferredPodPrimaryInterfaceName: special-nic
+```
+
+virt-handler and virt-launcher will be adjusted to prefer this value when trying to bind the primary network interface.
+They will both fallback to `eth0` in case they cannot find the preferred interface name.
+
+#### Pros
+
+- Does not require Multus to work
+
+#### Cons
+
+- Requires manual configuration
+- Error prone
+- Has cluster-wide granularity
+- Requires a change to the communication protocol between virt-handler and virt-launcher to pass the cluster-wide
+  setting to virt-launcher
+
+### Specifying the primary pod NIC name in advance
+
+When creating a VM with secondary networks, KubeVirt specifies the pod interfaces’ names for all secondary interfaces.
+It is currently not possible to specify the primary interface’s name when templating the virt-launcher pod, since there
+is no K8s API that enables it.
+
+The primary pod interface name is set by the container runtime (for example cri-o), before invoking the default CNI.
+
+#### Pros
+
+#### Cons
+
+- There is no K8s API to name the pod's primary NIC
+- There is no Multus API or config to name the pod's primary NIC
```

Comment:
It is a con or a no-go because of these points?

Reply:
Its a no-go.